ITEP-68793: Support Volumetric Analytics in Scene Controller (#148)
open-edge-platform/scenescape: docs/user-guide/How-to-configure-spatial-analytics.md

# How to Configure Spatial Analytics in Intel® SceneScape

> This guide provides step-by-step instructions to set up and use Regions Of Interest (ROIs) and Tripwires in Intel® SceneScape. By completing this guide, you will:

Reviewer: ```suggestion
This guide provides step-by-step instructions to set up and use Regions of Interest (ROIs) and Tripwires in Intel® SceneScape. By completing this guide, you will:
```

> - Understand the differences between regions of interest and tripwires

Reviewer: ```suggestion
- Understand the differences between Regions of Interest (ROIs) and Tripwires
```

- Learn how to configure ROIs and Tripwires through the UI
- Verify that events are properly triggered when objects interact with your defined analytics

---

## Prerequisites

> Follow the steps in [Getting-Started-Guide.md](./Getting-Started-Guide.md) to bringup an instance of SceneScape with out-of-box demo scenes.

Reviewer: ```suggestion
Follow the steps in [Getting-Started-Guide.md](./Getting-Started-Guide.md) to bring up an instance of SceneScape with out-of-box demo scenes.
```

## Steps to Configure Regions of Interest

### 1. Understand Analytic Types

**Regions of Interest (ROIs)** are defined areas within a scene where you want to monitor object presence, count, and dwell time.
**Tripwires** are virtual lines that trigger events when objects cross them in either direction.

---

### 2. Configure and Use a Region of Interest

#### Create a Region of Interest

1. Log in to Intel® SceneScape.
2. Click on a scene.
3. Click on `Regions` tab at the bottom of the page.
4. Click `New Region` button to create a region.

> 5. Fill out `Name`, `Description`, and select the `Scene`.

Reviewer: Remove this step -- it seems out of place and doesn't match the task flow.

> 6. Draw the region on the scene by clicking points to form a polygon.

Reviewer: ```suggestion
6. Draw the region on the scene by clicking points on the scene map to form a polygon. Be sure to click on the starting point to close the polygon.
```

7. **Optional**: Add a user-defined name for the ROI in the text box.

> 8. Click `Save Regions and Tripwires` to create the region.

Reviewer: ```suggestion
8. Click `Save Regions and Tripwires` to save the newly created region.
```

#### Modify a Region of Interest

1. Click on `Regions` at the bottom of the page.
2. Find your region in the Scene and double click on the polygon to edit its shape.
3. Click `Save Regions and Tripwires` to persist your changes.

#### Verify the Results

> 1. Use a tool like MQTT Explorer to monitor the topic right under the region name text box. ex: /scenescape//event/region/${scene_uuid}/${region_uuid}/count

Reviewer: ```suggestion
1. Use a tool like [MQTT Explorer](https://mqtt-explorer.com/) to monitor the topic right under the region name text box. ex: /scenescape//event/region/${scene_uuid}/${region_uuid}/count
```

> 2. When the center of the object enters/exits the region of interest, observe a message is received on that topic and it contains the following data:

Reviewer: ```suggestion
2. When the center of the object enters or exits the Region of Interest, observe that a message is received on the region event topic. Here is an example:
```

                    0,
                    0,
                    0,
                    1
                ],
                "visibility": [
                    "atag-qcam1",
                    "atag-qcam2"
                ],
                "regions": {},
                "similarity": null,
                "first_seen": "2025-07-11T06:29:02.378Z"
            },
            "dwell": 2.799999952316284 //What is the amount of time spent by the object in the ROI (in seconds)
        }
    ],
    "metadata": {
        "points": [
            [
                0.6242038216560509,
                4.617834394904459
            ],
            [
                1.7452229299363058,
                3.050955414012739
            ],
            [
                3.859872611464968,
                3.9426751592356686
            ],
            [
                2.2229299363057327,
                5.777070063694268
            ]
        ],
        "title": "roi_79c9e88c-6b26-482a-9a58-2f0c1b79bb05",
        "uuid": "79c9e88c-6b26-482a-9a58-2f0c1b79bb05",
        "area": "poly",
        "fromSensor": false
    }
}
```

![Configure and Verify Region of Interest](images/create-roi.gif)

> Figure 1: Region of interest creation flow

Reviewer: ```suggestion
Figure 1: Region of Interest creation flow
```

---

#### Enable Volumetric Intersection for Region of Interest

> Default behavior of Region of Interest is to trigger when the position of the object is within the bounds of the polygon. However, for detecting an event like a collision, computing a volumetric intersection is necessary.

Reviewer: ```suggestion
By default, Regions of Interest trigger events when the center point of each object enters or leaves the bounds of the polygon. However, for detecting an event like a collision, computing a volumetric intersection is necessary.
```

1. Follow the instructions in [How-to-define-object-properties.md](./How-to-define-object-properties.md) to create an entry for the object category of interest.

> 1. Click on `Regions` at the bottom of the page.

Reviewer: ```suggestion
1. Click on the `Regions` tab in the scene management page.
```

1. Find the specific region in the list and click on "volumetric" checkbox to enable intersection detection.
1. **Optional**: you can add a uniform buffer around the region and vary the height of the region.
1. Click `Save Regions and Tripwires` to persist your changes.

#### Verify the Results

> 1. Use a tool like MQTT Explorer to monitor the topic right under the region name text box. ex: /scenescape//event/region/${scene_uuid}/${region_uuid}/count

Reviewer: ```suggestion
1. Use a tool like MQTT Explorer to monitor the topic right under the region name text box. For example, `/scenescape/event/region/${scene_uuid}/${region_uuid}/count`
```

2. Navigate to the 3D UI view of the Scene.

> 3. When an object first intersects/last intersects with the region of interest, observe a message is received on that topic and it contains the following data:

Reviewer: ```suggestion
3. When an object first intersects/last intersects with the region of interest, observe a message is received on the event topic for that region. Here is an example:
```

                    0,
                    1
                ],
                "visibility": [
                    "atag-qcam1",
                    "atag-qcam2"
                ],
                "regions": {},
                "similarity": null,
                "first_seen": "2025-07-11T06:29:02.378Z"
            },
            "dwell": 2.799999952316284 //What is the amount of time spent by the object in the ROI (in seconds)
        }
    ],
    "metadata": {
        "points": [
            [
                0.6242038216560509,
                4.617834394904459
            ],
            [
                1.7452229299363058,
                3.050955414012739
            ],
            [
                3.859872611464968,
                3.9426751592356686
            ],
            [
                2.2229299363057327,
                5.777070063694268
            ]
        ],
        "title": "roi_79c9e88c-6b26-482a-9a58-2f0c1b79bb05",
        "uuid": "79c9e88c-6b26-482a-9a58-2f0c1b79bb05",
        "area": "poly",
        "fromSensor": false
    }
}
```

### 3. Configure and Use a Tripwire

#### Create a Tripwire

1. Log in to Intel® SceneScape.
2. Click on a scene.

> 3. Click on `Tripwires` at the bottom of the page.

Reviewer: ```suggestion
3. Click on the `Tripwires` tab below the scene map view.
```

4. Click `New Tripwire` to create a tripwire.
5. Click on the Scene and a green line with two moveable endpoints will appear.

> 6. Click and drag each endpoint to get the right orientation and position for the tripwire.

Reviewer: ```suggestion
6. Click and drag each endpoint to get the right orientation and position for the tripwire (the flag line indicates the direction of positive flow).
```

7. **Optional**: Add a user-defined name for the tripwire in the textbox
8. Click `Save Regions and Tripwires` to create the tripwire.

#### Modify a Tripwire

1. Click on `Tripwires` at the bottom of the page.
2. Double click on the tripwire to edit on the scene.
3. Click and drag to change position and orientation.
4. Click `Save Regions and Tripwires` to persist your changes.

#### Verify the Results

> 1. Use a tool like MQTT Explorer to monitor the topic right under the tripwire name text box. ex: /scenescape//event/tripwire/${scene_uuid}/${tripwire_uuid}/objects

Reviewer: ```suggestion
1. Use a tool like MQTT Explorer to monitor the topic right under the tripwire name text box. For example, `/scenescape/event/tripwire/${scene_uuid}/${tripwire_uuid}/objects`
```
